### PR TITLE
[#1861] Added 'org-oidc' form authentication backend using admins OpenID Connect config

### LINKED ISSUE #1861: Allow employees to login to a form using OpenID-connect
DH Taiga 95

**Tasks**

* [x] Add an Authentication plugin for OpenID-connect (this is VERY similar to: https://github.com/open-formulieren/open-forms/blob/master/src/openforms/authentication/contrib/digid_eherkenning_oidc/plugin.py)
* [x] Make sure the plugin re-uses the general OpenID-connect configuration (the one under the Account menu settings in the admin).
* [x] Make sure we can configure a claim that holds the employee-number (much like the kvk-number in above linked code).
* [x] Make sure the employee number is saved on the form submission (much like how the bsn, kvk, pseudo id are saved at the moment). It should be hashed (or whatever is used for bsn, etc.)
* [x] Test the system as a whole against Azure AD and as an employee-number claim we can use the email (configured as mapping in the admin), since I'm sure you will get that.

### PR DESCRIPTION
**Pull request by Bartvaderkin**
Fixes #1861

### CODE REVIEW

File: src/openforms/authentication/contrib/org_oidc/tests/test_auth_logout.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        session.save()
+
+        request = RequestFactory().post("/dummy")
+        request.session = session
+        request.user = user
+
+        plugin.logout(request)
+
+        self.assertEqual(list(request.session.keys()), [])
+        self.assertFalse(auth.get_user(self.client).is_authenticated)
```
Comment (sergei-maertens) [inline src/openforms/authentication/contrib/org_oidc/tests/test_auth_logout.py:59]: `auth.get_user` takes a `request` argument, but you seem to pass an instance of `django.test.Client` here. Am I missing something?

File: src/openforms/authentication/contrib/org_oidc/tests/test_auth_procedure.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+    oidc_op_user_endpoint="http://provider.com/auth/realms/master/protocol/openid-connect/userinfo",
+    username_claim="sub",
+    make_users_staff=True,
+)
+
+
+@override_settings(CORS_ALLOW_ALL_ORIGINS=True, IS_HTTPS=True)
+class OrgOIDCTests(TestCase):
+    @classmethod
+    def setUpTestData(cls):
```
Comment (sergei-maertens) [inline src/openforms/authentication/contrib/org_oidc/tests/test_auth_procedure.py:35]: don't forget the `super()` call!

File: src/openforms/authentication/contrib/org_oidc/urls.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+from django.urls import path
+
+from mozilla_django_oidc.urls import urlpatterns
+
+from openforms.authentication.contrib.org_oidc.views import (
```
Comment (sergei-maertens) [inline src/openforms/authentication/contrib/org_oidc/urls.py:5]:
```suggestion
from .views import (
```

File: src/openforms/authentication/contrib/org_oidc/tests/test_auth_logout.py (hunk 1 above)
Comment (Bartvaderkin) [reply src/openforms/authentication/contrib/org_oidc/tests/test_auth_logout.py:59]: The `auth.get_user()` only looks at `.session` so you can pass the `self.client` in tests.

File: src/openforms/authentication/contrib/org_oidc/tests/test_auth_logout.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        plugin = OIDCAuthentication("org-oidc")
+
+        request = RequestFactory().post("/dummy")
+        request.session = self.client.session
+        request.user = user
+
+        plugin.logout(request)
+
+        self.assertEqual(list(request.session.keys()), [])
+        self.assertFalse(auth.get_user(self.client).is_authenticated)
```
[the remark this answers is not in the document]
Comment (Bartvaderkin) [reply src/openforms/authentication/contrib/org_oidc/tests/test_auth_logout.py:32]: The auth.get_user() only looks at .session so you can pass the self.client in tests.

File: docs/configuration/authentication/oidc_org.rst [hunk 5]
```diff
@@ -0,0 +1,48 @@
+.. _configuration_authentication_oidc_org:
+
+=======================================
+OpenID Connect for organisation members
+=======================================
+
+Open Forms support login on forms by *organisation members* through Single Sign On (SSO) via the OpenID Connect protocol (OIDC).
```
Comment (sergei-maertens) [inline docs/configuration/authentication/oidc_org.rst:7]: I checked the "organization" spelling:

```
organization: 16 hits / 7 files
organisation: 7 hits / 7 files
```

-> so let's stick to `organization` as that's used most often. I've updated the other files in cd5d9155b1a66197e256be9f614b48aec3567409

Comment (sergei-maertens) [inline docs/configuration/authentication/oidc_org.rst:7]:
```suggestion
Open Forms supports login on forms by *organization members* through Single Sign On (SSO) via the OpenID Connect protocol (OIDC).
```

File: docs/configuration/authentication/oidc_org.rst [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+Members of the organisation can login to forms for internal use by the organisation using the same OpenID Connect configuration that is used for the management interface.
+In this flow:
+
+1. The organisation member clicks *Inloggen met OpenID Connect* on the internal form.
+2. The member gets redirected to the login flow of the OpenID Connect provider where the member logs in.
+3. After completion of the OpenID Connect provider login the member is redirected back to the form in Open Forms.
+4. The member is now logged into Open Forms and can proceed to complete the form.
+5. Open Forms optionally stores a configurable "Employee ID" claim from the OIDC user-info on both the submission as the created user.
```
Comment (sergei-maertens) [inline docs/configuration/authentication/oidc_org.rst:16]:
```suggestion
5. Open Forms optionally stores the employee ID (from the OIDC claims) on the user record _and_ the form submission.
```

File: docs/configuration/authentication/oidc_org.rst [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+==================================================
+
+The OpenID Connect configuration is shared with :ref:`the configuration of the management interface <configuration_authentication_oidc>` and follows the same steps with a few addtional notes:
+
+- The plugin callback URL needs to be registered at the OpenID Connect provider. Contact the organisations IAM and ask them add to the whitelist ``https://open-formulieren.gemeente.nl/org-oidc/callback/`` (replace ``open-formulieren.gemeente.nl`` with the domain of your Open Forms installation).
+
+- It is not recommended to use the *Default groups* configuration option when using OpenID Connect for organisation members to authenticate on forms.
+
+- To store user information from OpenID and track an "Employee ID" it is required to configure the `claim mapping`. This is JSON object where the claims from the OIDC user-info gets mapped to attributes on the user in Open Forms. For more info and options on configuring the mapping see `mozilla-django-oidc-db <https://github.com/maykinmedia/mozilla-django-oidc-db#user-content-user-profile>`_ and the documentation of your OpenID Connect provider for the structure of the returned user-info.
```
Comment (sergei-maertens) [inline docs/configuration/authentication/oidc_org.rst:29]:
```suggestion
- To store user information from OpenID and track an "Employee ID" it is required to configure the ``claim mapping``. This is JSON object where the claims from the OIDC user-info gets mapped to attributes on the user in Open Forms. For more info and options on configuring the mapping see `mozilla-django-oidc-db <https://github.com/maykinmedia/mozilla-django-oidc-db#user-content-user-profile>`_ and the documentation of your OpenID Connect provider for the structure of the returned user-info.
```

File: docs/configuration/authentication/oidc_org.rst [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+  Example:
+
+  .. code-block:: javascript
```
Comment (sergei-maertens) [inline docs/configuration/authentication/oidc_org.rst:33]:
```suggestion
  .. code-block:: json
```

I think this has its own directive.

File: docs/configuration/authentication/oidc_org.rst [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        "employee_id": "name_of_claim_with_employee_id"
+    }
+
+  Note we set the ``employee_id`` to track the member on both the submission and the created user.
+
+
+After completing these steps a form can be created with the authentication backend ``Organisation via OpenID Connect``, see :ref:`manual_forms_basics`.
+
+
+
```
Comment (sergei-maertens) [inline docs/configuration/authentication/oidc_org.rst:48]:
```suggestion
```

File: src/openforms/authentication/api/views.py [hunk 10]
```diff
@@ -12,6 +12,7 @@
 from openforms.submissions.utils import remove_submission_from_session
 from openforms.utils.api.views import ListMixin
 
+from ...api.authentication import AnonCSRFSessionAuthentication
```
Comment (sergei-maertens) [inline src/openforms/authentication/api/views.py:15]:
```suggestion
from openforms.api.authentication import AnonCSRFSessionAuthentication
```
really need to set up a linter/CI check for this :grimacing: 

this is a different "django app", so for those we use absolute imports

File: src/openforms/authentication/contrib/org_oidc/backends.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    OIDCAuthenticationBackend as _OIDCAuthenticationBackend,
+)
+
+from .mixins import SoloConfigMixin
+
+logger = logging.getLogger(__name__)
+
+
+class OIDCAuthenticationBackend(SoloConfigMixin, _OIDCAuthenticationBackend):
+    config_identifier_field = "username_claim"
```
Comment (sergei-maertens) [inline src/openforms/authentication/contrib/org_oidc/backends.py:13]: I can't immediately tell if this is used / different from the default - is this still needed?

File: src/openforms/authentication/contrib/org_oidc/backends.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
+import logging
+
+from mozilla_django_oidc_db.backends import (
+    OIDCAuthenticationBackend as _OIDCAuthenticationBackend,
+)
+
+from .mixins import SoloConfigMixin
+
+logger = logging.getLogger(__name__)
```
Comment (sergei-maertens) [inline src/openforms/authentication/contrib/org_oidc/backends.py:9]: unused

File: src/openforms/authentication/contrib/org_oidc/tests/test_auth_procedure.py [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+        super()
```
Comment (sergei-maertens) [inline src/openforms/authentication/contrib/org_oidc/tests/test_auth_procedure.py:36]:
```suggestion
        super().setUpTestData()
```

**Review by sergei-maertens — CHANGES_REQUESTED** [review]

File: src/openforms/authentication/api/views.py (hunk 10 above)
Comment (Bartvaderkin) [reply src/openforms/authentication/api/views.py:15]: Some linting/autofixing would be handy, I can't configure PyCharm's auto-import to make the distinction so these keep showing up everywhere.

**Review by sergei-maertens — APPROVED** [review]

**Pull request merged**